### DISCUSSION TITLE
Best way to collapse nested language

### DISCUSSION THREAD

**Discussion by genos** [egraphs-good/egg#260] [Q&A] [answered]
Hello all! 

First off, thanks _very_ much for `egg`, it's really interesting.

I'm trying to use `egg` to, among other things, flatten out a potentially nested structure, but I'm having trouble figuring out the best way to do it. Simplifying greatly from my intended application, suppose we've got the following language.

```rust
define_language! {
    pub enum Nested {
        One(Symbol),
        "many" = Many(Vec<Id>),
    }
}
```

What's the best way to flatten a potential multi-layered `Many`?[^1] Would it go in an `Analysis`, or in rewrite rules? For example, I'd ultimately like something like

```
(many (many (many 0)) 1 2 (many 3) (many (many 4) 5 (many 6 7 8)))  ⇒ (many 0 1 2 3 4 5 6 7 8)
```

But I'm having trouble figuring out how to pull this off:
- I'd imagine the ideal place to do this would be in an `Analysis::make`, but I can't see how to do that without cloning a bunch?
- I haven't been able to do it via `rewrite!`, as I'm not sure how to match on the vector of Ids instead of pattern matching on a singleton

Any guidance would be greatly appreciated. Thanks again!

[^1]: I'm thinking of something similar to the `Foldable` instance for [rose trees in Haskell](https://hackage.haskell.org/package/containers-0.6.7/docs/Data-Tree.html) if that's familiar to anyone.

**Comment by mwillsey** [comment, marked as answer]
Hi there!

Yes, that is definitely one of the downsides of the variadic patterns: you can't write a pattern than captures all of them. You could maybe try refactoring to a sort of "cons list" approach, but I'm not sure if that would work in your domain. I think you could definitely do it in an analysis. I wouldn't fear cloning tooooo much, after all, you have to materialize the list somewhere. Perhaps a clever use of Rc or something a [persistent vector](https://docs.rs/im/latest/im/struct.Vector.html) could help.

**Reply by genos** [reply]
Thanks for getting back to me so quickly! The cons list and immutable structures are nice ideas, thanks.

What I'm actually trying to do (which the above example grossly oversimplifies) is rewrite programs of [Quantum gates](https://en.wikipedia.org/wiki/Quantum_logic_gate), something like:

```rust
egg::define_language! {
    pub enum GateSet {
        Angle(OrderedFloat(f32)),
        "+" = Add([Id; 2]),
        "RX" = RX([Id; 2]),
        "RZ" = RZ([Id; 2]),
        "H" = H(Id),
        "circuit" = Circuit(Vec<Id>),
        Symbol(egg::Symbol),
    }
}

```
Here a single gate might first be expanded to a collection of multiple gates applied in a row; for example, the hardware involved might decompose a [Hadamard gate](https://en.wikipedia.org/wiki/Quantum_logic_gate#Hadamard_gate) as a series of rotations:

```rust
rw!("h"; "(H ?q)" => "(circuit (RZ pi/2 ?q) (RX pi/2 ?q) (RZ pi/2 ?q))")
```
I'd also like the entire collection gets reduced by a number of rules, like "two consecutive `RZ`s on the same qubit can be simplified to a single `RZ` with the sum of their angles." like

```rust
rw!("rz-repeated"; "(circuit (RZ ?a1 ?q) (RZ ?a2 ?q))" => "(circuit (RZ (+ ?a1 ?a2) ?q))")
```

Rubber-ducking it a bit, would it make sense to split the stages in two? Handle the expansion to a native gate set separately, then use `egg` to simplify the larger chain, perhaps in a cons list structure; the former is a bit finicky/hardware specific, while the latter bit is essentially just group theory.

**Reply by mwillsey** [reply]
Hmm, I'm not sure, potentially! I'm not really sure what a "circuit" is in this context; this is definitely outside my wheelhouse. If it really is just a list of things, and you only need to match on adjacent neighbors, I think cons list might really be helpful here.

**Reply by genos** [reply]
Oh! Since “circuits” here are lists of instructions to be applied in sequential order and gate operations form a group (sweeping details under the rug), I could have a language entry 
```rust
“seq” = Seq([Id; 2])
``` 
and add associativity as part of the rewrite rules.

…or was that what you had in mind when mentioning cons lists 😅

Thanks so much again for your help and for the library!

**Comment by mwillsey** [comment]
Sure thing! One note: I would watch out for associativity, as it can
quickly explode the e-graph if you aren't careful. Maybe your usecase will
just work; but it's something to watch out for!
